File: concurrent-programming/cpp/thread_pool.cpp
// #include <iostream>
// #include <thread>
// #include <vector>
// #include <queue>
// #include <functional>
// #include <mutex>
// #include <condition_variable>
// #include <chrono>
// #include <atomic>

// std::mutex coutMtx;

// using Task = std::function<void()>;

// class ThreadPool {
// public:
//     ThreadPool(size_t numThreads) : mb_Stop(false) {
//         for (size_t i = 0; i < numThreads; ++i) {
//             m_Workers.emplace_back([this] { workerThread(); });
//         }
//     }

//     ~ThreadPool() {
//         {
//             std::unique_lock<std::mutex> lock(m_queueMtx);
//             mb_Stop = true;
//         }
//         m_CV.notify_all();
//         for (auto& worker : m_Workers) {
//             worker.join();
//         }
//     }

//     template <typename Func, typename... Args>
//     void Submit(Func&& func, Args&&... args) {
//         {
//             std::unique_lock<std::mutex> lk(m_queueMtx);
//             m_Tasks.emplace(std::bind(std::forward<Func>(func), std::forward<Args>(args)...));
//         }
//         m_CV.notify_one();
//     }

//     void Wait() {
//         std::unique_lock<std::mutex> lk(m_WaitMtx);
//         m_CVWait.wait(lk, [this] { return m_Tasks.empty(); });
//     }

// private:
//     void workerThread() {
//         while (true) {
//             Task task;
//             {
//                 std::unique_lock<std::mutex> lock(m_queueMtx);
//                 m_CV.wait(lock, [this] { return mb_Stop || !m_Tasks.empty(); });
//                 if (mb_Stop && m_Tasks.empty()) {
//                     return;
//                 }
//                 task = std::move(m_Tasks.front());
//                 m_Tasks.pop();

//                 if(m_Tasks.empty()){
//                     m_CVWait.notify_one();
//                 }
//             }

//             task();
//         }
//     }

//     std::vector<std::thread> m_Workers;
//     std::queue<Task> m_Tasks;
//     std::mutex m_queueMtx;
//     std::condition_variable m_CV;
//     std::mutex m_WaitMtx;
//     std::condition_variable m_CVWait;
//     std::atomic<bool> mb_Stop;
// };

// void cpu_waster(int idx) {
//     std::this_thread::sleep_for(std::chrono::seconds(3));
//     {   
//         std::lock_guard<std::mutex> lk(coutMtx);
//         std::cout << std::this_thread::get_id() << " : doing " << idx << " work" << std::endl;
//     }
// }

// int main() {
//     int numJobs = 20;
//     int numThreads = 5;

//     ThreadPool pool(numThreads);

//     for (int idx = 0; idx < numJobs; ++idx) {
//         pool.Submit(cpu_waster, idx);
//     }
//     pool.Wait();

//     return 0;
// }

File: concurrent-programming/python/patterns/thread_pool.py
import time
from queue import Queue
import threading
from threading import Thread
from typing import Callable, Any, Tuple

from utils import Timer

# typedef
Callback = Callable[..., None]
Task = Tuple[Callback, Any, Any]

class Worker(Thread):
    def __init__(self, tasks:Queue[Task]):
        super().__init__()
        self.tasks : Queue = tasks
    
    def run(self):
        while True:
            if not self.tasks.empty():
                func, args, kwargs = self.tasks.get()
                func(*args, **kwargs)
                self.tasks.task_done()

class ThreadPool:
    def __init__(self, num_threads:int):
        self.tasks = Queue(num_threads)
        self.num_threads = num_threads

        for _ in range(self.num_threads):
            worker = Worker(self.tasks)
            worker.setDaemon(True)
            worker.start()

    def submit(self, func:Callback, *args, **kwargs):
        self.tasks.put((func, args, kwargs))
    
    def wait(self):
        self.tasks.join()

def cpu_waster(idx: int) -> None:
    name = threading.current_thread().getName()
    print(f"{name} : doing {idx} work")
    time.sleep(3)

def main():
    num_jobs     : int   = 20
    num_threads  : int   = 5
    timer        : Timer = Timer()

    pool = ThreadPool(num_threads = num_threads)
    for idx in range(num_jobs):
        pool.submit(cpu_waster, idx)
    pool.wait()

if "__main__" == __name__:
    main()

